frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,828 @@
+				None
+			}
+			Err((score, insert_idx)) => {
+				// could not insert into the set because it is full.
+				// note that we short-circuit return here in case the iteration produces `None`.
+				// If there wasn't a weakest entry to remove, then there must be a capacity of 0,
+				// which means that we can't meaningfully proceed.
+				let (weakest_score, weakest_idx) = match self.indices.iter().next() {
+					None => return (false, None),
+					Some((score, idx)) => (*score, *idx),
+				};
+				let threshold = T::SolutionImprovementThreshold::get();
+
+				// if we haven't improved on the weakest score, don't change anything.
+				if !is_score_better(score, weakest_score, threshold) {
+					return (false, None);
+				}
+
+				self.indices.remove(&weakest_score);
+				self.indices
+					.try_insert(score, insert_idx)
+					.expect("just removed an item, we must be under capacity; qed");
+
+				// ensure that SignedSubmissionsMap never grows past capacity by taking out the
+				// weakest member here.
+				self.take_submission(weakest_idx)
+			}
+		};
+
+		// we've taken out the weakest, so update the storage map and the next index
+		self.insertion_overlay.insert(self.next_idx, submission);
+		debug_assert!(!self.deletion_overlay.contains(&self.next_idx));
+		self.next_idx += 1;
+		(true, weakest)
+	}
+
+	/// Remove the signed submission with the highest score from the set.
+	pub fn pop_last(&mut self) -> Option<SignedSubmissionOf<T>> {
+		let (highest_score, idx) = self.indices.iter().rev().next()?;
+		let (highest_score, idx) = (*highest_score, *idx);
+		self.indices.remove(&highest_score);
+		self.take_submission(idx)
+	}
+}
+
+impl<T: Config> Deref for SignedSubmissions<T> {
+	type Target = SubmissionIndicesOf<T>;
+
+	fn deref(&self) -> &Self::Target {
+		&self.indices
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// `Self` accessor for `SignedSubmission<T>`.
+	pub fn signed_submissions() -> SignedSubmissions<T> {
+		SignedSubmissions::<T>::get()
+	}
+
+	/// Finish the signed phase. Process the signed submissions from best to worse until a valid one
+	/// is found, rewarding the best one and slashing the invalid ones along the way.
+	///
+	/// Returns true if we have a good solution in the signed phase.
+	///
+	/// This drains the [`SignedSubmissions`], potentially storing the best valid one in
+	/// [`QueuedSolution`].
+	pub fn finalize_signed_phase() -> (bool, Weight) {
+		let mut all_submissions = Self::signed_submissions();
+		let mut found_solution = false;
+		let mut weight = T::DbWeight::get().reads(1);
+
+		while let Some(best) = all_submissions.pop_last() {
+			let SignedSubmission { solution, who, deposit, reward } = best;
+			let active_voters = solution.compact.voter_count() as u32;
+			let feasibility_weight = {
+				// defensive only: at the end of signed phase, snapshot will exits.
+				let SolutionOrSnapshotSize { voters, targets } =
+					Self::snapshot_metadata().unwrap_or_default();
```

Comment:
Can we read the `snapshot_metadata` outside of this loop? From what I can tell it is invariant for the duration of this while loop

Reply:
Would be a nice micro-opt 👍 albeit it is always cached, so it is not horrible now either.

Reply:
2355a7844

Reply:
https://github.com/paritytech/substrate/pull/7910/commits/2355a784451b4c59b9c2ef4f1ec22ae8f04db4d1